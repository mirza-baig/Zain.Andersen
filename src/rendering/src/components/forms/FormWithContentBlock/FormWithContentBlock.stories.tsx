// Global
import { SitecoreContext } from '@sitecore-jss/sitecore-jss-react';
import { Story, Meta } from '@storybook/react';
import { createComponentFactory } from 'lib/mocks/mock-placeholder';
// Local
import FormWithContentBlock, { FormWithContentBlockProps } from './FormWithContentBlock';
import defaultData from './FormWithContentBlock.mock-data';

const componentFactory = createComponentFactory();
export default {
  title: 'Forms/FormWithContentBlock',
  component: FormWithContentBlock,
  decorators: [
    (Story) => <SitecoreContext componentFactory={componentFactory}>{Story()}</SitecoreContext>,
  ],
} as Meta;
const Template: Story<FormWithContentBlockProps> = (props) => <FormWithContentBlock {...props} />;
export const Default = Template.bind({});
Default.args = defaultData;
