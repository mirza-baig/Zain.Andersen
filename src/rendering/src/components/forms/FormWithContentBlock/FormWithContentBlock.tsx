import { withDatasourceCheck, Placeholder } from '@sitecore-jss/sitecore-jss-nextjs';
import { Feature } from 'src/.generated/Feature.EnterpriseWeb.model';
import classNames from 'classnames';
import { getEnum } from 'lib/utils';
import { Component } from 'src/helpers/Component';
import { ComponentBackgroundVariants } from 'src/helpers/Component/Component';

export type FormWithContentBlockProps =
  Feature.EnterpriseWeb.RenewalByAndersen.Components.Forms.FormWithContentBlock.FormWithContentBlock;

const FormWithContentBlock = (props: FormWithContentBlockProps) => {
  const { rendering, fields } = props;
  const style = getEnum<ComponentBackgroundVariants>(fields?.backgroundColor) ?? 'white';
  const isContentBlockPresent = rendering?.placeholders?.contentBlock?.length > 0;
  const formColClasses = classNames(
    'col-span-12',
    isContentBlockPresent
      ? 'max-md:mx-m max-md:border-b max-md:border-solid max-md:border-gray md:col-span-5'
      : 'max-md:px-m  md:col-span-4 md:col-start-5 '
  );

  return (
    <Component
      variant="full"
      gap="gap-x-0"
      padding="px-0"
      backgroundVariant={style}
      dataComponent="forms/formwithcontentblock"
      {...props}
    >
      <div className="col-span-12 py-l">
        <div className="grid grid-cols-2 gap-x-0 max-lg:px-m max-md:px-0 md:max-w-screen-lg md:grid-cols-12 lg:mx-auto">
          <div className={formColClasses}>
            <Placeholder name="form" rendering={rendering} />
          </div>
          {isContentBlockPresent && (
            <div className="col-span-12 md:col-span-6 md:col-start-7 md:my-2 md:border-l md:border-solid md:border-gray">
              <Placeholder name="contentBlock" rendering={rendering} />
            </div>
          )}
        </div>
      </div>
    </Component>
  );
};

export default withDatasourceCheck()<FormWithContentBlockProps>(FormWithContentBlock);
