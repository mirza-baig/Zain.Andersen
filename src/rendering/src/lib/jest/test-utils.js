// Global
import { render, fireEvent, waitFor } from '@testing-library/react';
import { ComponentPropsContext } from '@sitecore-jss/sitecore-jss-nextjs';
import { FastSitecoreContext } from 'lib/overrides/FastSitecoreContext';

// Lib
import { createSampleComponentFactory } from 'lib/mocks/mock-placeholder';

export const eeSitecoreContext = {
  route: {
    name: 'Foo',
    placeholders: {},
  },
  context: {
    pageEditing: true,
    site: {
      name: 'JssTestWeb',
    },
    language: 'en',
  },
};

export const renderComponent = (
  Component,
  {
    componentProps = {},
    staticProps = {},
    componentFactory = createSampleComponentFactory(),
    sitecoreContext = {
      route: {
        name: 'Foo',
        placeholders: {},
      },
    },
  }
) => {
  const mockLayoutData = {
    sitecore: sitecoreContext,
  };

  return render(
    <ComponentPropsContext value={staticProps}>
      <FastSitecoreContext componentFactory={componentFactory} layoutData={mockLayoutData}>
        <Component {...componentProps} />
      </FastSitecoreContext>
    </ComponentPropsContext>
  );
};

export const snapshot = (Component, mockData) => {
  const component = renderComponent(Component, mockData);
  expect(component.container.firstChild).toMatchSnapshot();
  return component;
};

export const hasDataComponent = async (component, name) => {
  await waitFor(() => {
    const nodes = component.baseElement.querySelectorAll(`[data-component="${name}"]`);
    return expect(nodes.length).toBe(1);
  });
};

export const mockDateAs = (epochMs) => {
  if (typeof epochMs !== 'number') {
    console.error(
      'mockDateAs only accepts epoch values in milliseconds to prevent timezones from factoring in to Date creation.'
    );
  }

  jest.spyOn(global.Date, 'now').mockImplementation(() => new Date(epochMs).valueOf());
};

export const testFieldError = async (Component, mockData, { label, value, error }) => {
  const component = renderComponent(Component, mockData);
  const field = component.getByLabelText(label);
  fireEvent.change(field, { target: { value } });
  field.blur();
  await waitFor(() => {
    const err = component.queryByText(error);
    return expect(err).not.toBe(null);
  });
};
