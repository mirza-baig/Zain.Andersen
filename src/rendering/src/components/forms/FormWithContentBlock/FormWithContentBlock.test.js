// Global
import { hasDataComponent, snapshot } from 'lib/jest/test-utils';
// Local
import FormWithContentBlock from './FormWithContentBlock';
import defaultData from './FormWithContentBlock.mock-data';

it('renders correctly', async () => {
  const component = snapshot(FormWithContentBlock, { componentProps: defaultData });
  await hasDataComponent(component, 'forms/formwithcontentblock');
});
