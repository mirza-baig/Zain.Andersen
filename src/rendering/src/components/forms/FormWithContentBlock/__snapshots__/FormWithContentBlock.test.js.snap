// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<section
  data-component="forms/formwithcontentblock"
  id="id2142262536"
>
  <div
    class="my-4 w-full theme-gray bg-light-gray"
  >
    <div
      class="section-grid grid grid-cols-2 md:grid-cols-12 px-0 gap-x-0 "
    >
      <div
        class="col-span-12 py-l"
      >
        <div
          class="grid grid-cols-2 gap-x-0 max-lg:px-m max-md:px-0 md:max-w-screen-lg md:grid-cols-12 lg:mx-auto"
        >
          <div
            class="col-span-12 max-md:mx-m max-md:border-b max-md:border-solid max-md:border-gray md:col-span-5"
          >
            <div
              style="background: darkorange; outline: 5px solid orange; padding: 10px; color: white; max-width: 500px;"
            >
              <h2>
                Form
              </h2>
              <p>
                JSS component is missing React implementation. See the developer console for more information.
              </p>
            </div>
          </div>
          <div
            class="col-span-12 md:col-span-6 md:col-start-7 md:my-2 md:border-l md:border-solid md:border-gray"
          >
            <div
              style="background: darkorange; outline: 5px solid orange; padding: 10px; color: white; max-width: 500px;"
            >
              <h2>
                ContentBlock
              </h2>
              <p>
                JSS component is missing React implementation. See the developer console for more information.
              </p>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</section>
`;
